---
title: "Rachel_QTL_GWAS"
author: "Rachel Olson"
date: "March 7, 2020"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, echo=FALSE}
#libraries
library(qtl);set.seed(043298)
library("tidyverse")

#load
load("~/R/Classes/PLS298_QuantGen/PLB298FINAL/FinalProjectData.RData")
markers <- read.table("~/R/Classes/PLS298_QuantGen/PLB298FINAL/markers.txt", quote="\"", comment.char="")
```

```{r, echo=FALSE}
#Add Chromosome and position rows to G
colnames(markers) <- c("Chr","Position","Marker","A","B"); tmarkers <- t(markers); colnames(tmarkers) <- markers$Marker; tmarkerchr <- data.frame(tmarkers["Chr",], tmarkers["Position",]); colnames(tmarkerchr) <- c("Chr", "Position"); markerchr <- t(tmarkerchr)
G <- rbind(markerchr, G)
rownames(G)[rownames(G) == "Chr"] = ""
rownames(G)[rownames(G) == "Position"] = ""

#convert homologous chromosomes to functionally single chromosome
G <- gsub("Fvb","",G)
#because I can't find an elegant way to do this:
G <- gsub("1-1","1",G);G <- gsub("1-2","1",G);G <- gsub("1-3","1",G);G <- gsub("1-4","1",G);G <- gsub("6-2","6",G);G <- gsub("2-1","2",G);G <- gsub("2-2","2",G);G <- gsub("2-3","2",G);G <- gsub("2-4","2",G);G <- gsub("4-1","4",G);G <- gsub("3-3","3",G);G <- gsub("7-4","7",G);G <- gsub("7-1","7",G);G <- gsub("3-2","3",G);G <- gsub("3-1","3",G);G <- gsub("3-4","3",G);G <- gsub("6-4","6",G);G <- gsub("6-1","6",G);G <- gsub("4-3","4",G);G <- gsub("7-3","7",G);G <- gsub("4-2","4",G);G <- gsub("5-2","5",G);G <- gsub("5-3","5",G);G <- gsub("4-4","4",G);G <- gsub("5-4","5",G);G <- gsub("5-1","5",G);G <- gsub("7-2","7",G);G <- gsub("6-3","6",G)

#small edits to pheno data
colnames(pheno)[colnames(pheno) == "Progeny"] = "ID"
pheno <- pheno[,-5] #remove MotherID
pheno <- pheno[,-5] #remove FatherID
pheno[nrow(pheno) +1, ] <- NA #add missing parent to phenotype file
pheno$Progeny[nrow(pheno)] <- "inQ5IDMeqY"
G <- G[order(match(rownames(G),pheno$ID)),] # Reorder genotype dataframe to match

#put markers in order
tG <- t(G)
tG <- tG[order(tG[,1],tG[,2]),]
tG <- tG[,-2]
G <- t(tG)
ID <- rownames(G)
G <-cbind(ID,G)
rownames(G) <- NULL


#Write out files
write.csv(G, file = "G.csv")
write.csv(pheno, file = "pheno.csv")
```

```{r, echo=FALSE}
#Add cross
crossobj <- read.cross(c("csvs"), dir = "~/R/Classes/PLS298_QuantGen/PLB298FINAL/", genfile = "G.csv", phefile = "pheno.csv", estimate.map = FALSE, genotypes = c(-1,0,1,2,3), na.strings = c("-1",NA))
```

QTL Mapping and Linkage

```{r, echo=FALSE}
```

GWAS

```{r, echo=FALSE}

```


